### ISSUE TITLE
Auto num_worker tuning

### ISSUE THREAD

**Issue by GeorgePearse** [Lightning-AI/lightning#14961]
## 🚀 Feature
```
trainer = Trainer(...,
     num_worker_tune=True
)
```

### Motivation

After you've optimized for things that impact model training performance (batch_size and learning_rate), e.g. the things that you can reason about after an epoch or two and which PL provides good tooling to do, it makes sense to try to match other parameters to the machine you're training on with respect to training times. num_workers can make a big difference to loading times but isn't altogether simple to reason about.

### Pitch

Sweep through num_workers with time to load as the metric to optimise, start by matching CPU count, maybe just increase from there? Would experiment around a bit 

Would love to investigate / scope it, run some experiments and ideally implement it, but will need quite a lot of guidance.

### Additional Context

I guess unlike the batch_size and learning_rate it's a property of the data_module, not clear to me how much more awkward that makes it.

**Comment by carmocca** [comment]
This is a duplicate of https://github.com/Lightning-AI/lightning/issues/2196

It's unimplemented because there's no straightforward way to test what's the optimal num workers value. Any heuristic will depend on your hardware and training procedure, and they aren't necessarily more reliable than our current suggestions in warnings
